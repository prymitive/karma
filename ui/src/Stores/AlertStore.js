import { observable, action, computed, toJS } from "mobx";

import throttle from "lodash.throttle";

import qs from "qs";

import { FetchGet } from "Common/Fetch";

const QueryStringEncodeOptions = {
  encodeValuesOnly: true, // don't encode q[]
  indices: false, // go-gin doesn't support parsing q[0]=foo&q[1]=bar
};

function FormatAlertsQ(filters) {
  return qs.stringify({ q: filters }, QueryStringEncodeOptions);
}

// generate URL for the UI with a set of filters
function FormatAPIFilterQuery(filters) {
  return qs.stringify(
    Object.assign(DecodeLocationSearch(window.location.search).params, {
      q: filters,
    }),
    QueryStringEncodeOptions
  );
}

// format URI for react UI -> Go backend requests
function FormatBackendURI(path) {
  const uri = process.env.REACT_APP_BACKEND_URI || ".";
  return `${uri}/${path}`;
}

// takes the '?foo=bar&foo=baz' part of http://example.com?foo=bar&foo=baz
// and decodes it into a dict with some extra metadata
function DecodeLocationSearch(searchString) {
  let defaultsUsed = true;
  let params = { q: [] };

  if (searchString !== "") {
    const parsed = qs.parse(searchString.split("?")[1]);
    params = Object.assign(params, parsed);

    if (parsed.q !== undefined) {
      defaultsUsed = false;
      if (parsed.q === "") {
        params.q = [];
      } else if (Array.isArray(parsed.q)) {
        // first filter out duplicates
        // then filter out empty strings, so 'q=' doesn't end up [""] but rather []
        params.q = parsed.q
          .filter((v, i) => parsed.q.indexOf(v) === i)
          .filter((v) => v !== "");
      } else {
        params.q = [parsed.q];
      }
    }
  }

  return { params: params, defaultsUsed: defaultsUsed };
}

function UpdateLocationSearch(newParams) {
  const baseURLWithoutSearch = window.location.href.split("?")[0];
  const newSearch = FormatAPIFilterQuery(newParams.q);
  window.history.pushState(
    null,
    null,
    `${baseURLWithoutSearch}?${newSearch || "q="}`
  );
}

const AlertStoreStatuses = Object.freeze({
  Idle: Symbol("idle"),
  Fetching: Symbol("fetching"),
  Processing: Symbol("processing"),
  Failure: Symbol("failure"),
});

function NewUnappliedFilter(raw) {
  return {
    applied: false,
    isValid: true,
    raw: raw,
    hits: 0,
    name: "",
    matcher: "",
    value: "",
  };
}

class AlertStore {
  filters = observable(
    {
      values: [],
      addFilter(raw) {
        if (this.values.filter((f) => f.raw === raw).length === 0) {
          this.values.push(NewUnappliedFilter(raw));
          UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
        }
      },
      removeFilter(raw) {
        if (this.values.filter((f) => f.raw === raw).length > 0) {
          this.values = this.values.filter((f) => f.raw !== raw);
          UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
        }
      },
      replaceFilter(oldRaw, newRaw) {
        const index = this.values.findIndex((e) => e.raw === oldRaw);
        if (index >= 0) {
          // first check if we would create a duplicated filter
          if (this.values.findIndex((e) => e.raw === newRaw) >= 0) {
            // we already have newRaw, simply drop oldRaw
            this.removeFilter(oldRaw);
          } else {
            // no dups, continue with a swap
            this.values[index] = NewUnappliedFilter(newRaw);
            UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
          }
        }
      },
      setFilters(raws) {
        this.values = raws.map((raw) => NewUnappliedFilter(raw));
        UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
      },
      setWithoutLocation(raws) {
        const filtersByRaw = this.values.reduce(function (map, obj) {
          map[toJS(obj.raw)] = toJS(obj);
          return map;
        }, {});
        this.values = raws.map((raw) =>
          filtersByRaw[raw] ? filtersByRaw[raw] : NewUnappliedFilter(raw)
        );
      },
      applyAllFilters() {
        for (let i = 0; i < this.values.length; i++) {
          this.values[i].applied = true;
        }
      },
    },
    {
      addFilter: action.bound,
      removeFilter: action.bound,
      replaceFilter: action.bound,
      setFilters: action.bound,
      setWithoutLocation: action.bound,
      applyAllFilters: action.bound,
    },
    { name: "API Filters" }
  );

  data = observable(
    {
      colors: {},
      counters: [],
      grids: [],
      silences: {},
      upstreams: { instances: [], clusters: {} },
      receivers: [],
      getAlertmanagerByName(name) {
        return this.upstreams.instances.find((am) => am.name === name);
      },
      isReadOnlyAlertmanager(name) {
        return this.readOnlyAlertmanagers.map((am) => am.name).includes(name);
      },
      getClusterAlertmanagersWithoutReadOnly(clusterID) {
        return this.clustersWithoutReadOnly[clusterID] || [];
      },
      get readOnlyAlertmanagers() {
        return this.upstreams.instances.filter((am) => am.readonly === true);
      },
      get readWriteAlertmanagers() {
        return this.upstreams.instances
          .filter((am) => am.readonly === false)
          .map((am) =>
            Object.assign({}, am, {
              clusterMembers: am.clusterMembers.filter(
                (m) => this.isReadOnlyAlertmanager(m) === false
              ),
            })
          );
      },
      get clustersWithoutReadOnly() {
        const clusters = {};
        for (const clusterID of Object.keys(this.upstreams.clusters)) {
          const members = this.upstreams.clusters[clusterID].filter(
            (member) => this.isReadOnlyAlertmanager(member) === false
          );
          if (members.length > 0) {
            clusters[clusterID] = members;
          }
        }
        return clusters;
      },
      getColorData(name, value) {
        if (this.colors[name] !== undefined) {
          return this.colors[name][value];
        }
      },
    },
    {
      readOnlyAlertmanagers: computed,
      readWriteAlertmanagers: computed,
      clustersWithoutReadOnly: computed,
    },
    { name: "API Response data" }
  );

  info = observable(
    {
      authentication: {
        enabled: false,
        username: "",
      },
      totalAlerts: 0,
      version: "unknown",
      upgradeNeeded: false,
      isRetrying: false,
      reloadNeeded: false,
      setIsRetrying() {
        this.isRetrying = true;
      },
      clearIsRetrying() {
        this.isRetrying = false;
      },
      setReloadNeeded() {
        this.reloadNeeded = true;
      },
    },
    {
      setIsRetrying: action.bound,
      clearIsRetrying: action.bound,
      setReloadNeeded: action,
    },
    { name: "API response info" }
  );

  settings = observable(
    {
      values: {
        staticColorLabels: [],
        annotationsDefaultHidden: false,
        annotationsHidden: [],
        annotationsVisible: [],
        sorting: {
          grid: {
            order: "startsAt",
            reverse: false,
            label: "alertname",
          },
          valueMapping: {},
        },
        silenceForm: {
          strip: {
            labels: [],
          },
        },
        alertAcknowledgement: {
          enabled: false,
          durationSeconds: 900,
          author: "karma / author missing",
          commentPrefix: "",
        },
      },
    },
    {},
    {
      name: "Global settings",
    }
  );

  status = observable(
    {
      value: AlertStoreStatuses.Idle,
      lastUpdateAt: 0,
      error: null,
      paused: false,
      setIdle() {
        this.value = AlertStoreStatuses.Idle;
        this.error = null;
        this.lastUpdateAt = new Date();
      },
      setFetching() {
        this.value = AlertStoreStatuses.Fetching;
      },
      setProcessing() {
        this.value = AlertStoreStatuses.Processing;
        this.error = null;
      },
      setFailure(err) {
        this.value = AlertStoreStatuses.Failure;
        this.error = err;
        this.lastUpdateAt = new Date();
      },
      pause() {
        this.paused = true;
      },
      resume() {
        this.paused = false;
      },
    },
    {
      setIdle: action,
      setFetching: action,
      setProcessing: action,
      setFailure: action,
      pause: action.bound,
      resume: action.bound,
    },
    { name: "Store status" }
  );

  constructor(initialFilters) {
    if (initialFilters !== null) this.filters.setFilters(initialFilters);
  }

  fetch = action(
    (gridLabel, gridSortReverse, sortOrder, sortLabel, sortReverse) => {
      this.status.setFetching();

      const args = [
        `gridLabel=${gridLabel}`,
        `gridSortReverse=${gridSortReverse ? "1" : "0"}`,
        `sortOrder=${sortOrder}`,
        `sortLabel=${sortLabel}`,
        `sortReverse=${sortReverse}`,
      ];

      const alertsURI =
        FormatBackendURI(`alerts.json?&${args.join("&")}&`) +
        FormatAPIFilterQuery(this.filters.values.map((f) => f.raw));

      return FetchGet(alertsURI, {}, this.info.setIsRetrying)
        .then((result) => {
          // we're sending requests with mode=cors so the response should also be type=cors
          // after a few failures in the retry loop we will switch to no-cors
          // if that request comes back as type=opaque then we might be getting
          // redirected by an auth proxy
          if (result.type === "opaque") {
            this.info.setReloadNeeded();
          }
          this.info.clearIsRetrying();
          this.status.setProcessing();
          return result.json();
        })
        .then((result) => {
          return this.parseAPIResponse(result);
        })
        .catch((err) => {
          console.trace(err);
          return this.handleFetchError(
            `Can't connect to the API, last error was "${err.message}"`
          );
        });
    }
  );

  fetchWithThrottle = throttle(this.fetch, 300);

  parseAPIResponse = action((result) => {
    if (result.error) {
      this.handleFetchError(result.error);
      return;
    }

    const queryFilters = [
      ...new Set(
        this.filters.values
          .map((f) => f.raw)
          .slice()
          .sort()
      ),
    ];
    const responseFilters = [
      ...new Set(result.filters.map((m) => m.text).sort()),
    ];
    if (JSON.stringify(queryFilters) !== JSON.stringify(responseFilters)) {
      console.info(
        `Got response with filters '${responseFilters}' while expecting results for '${queryFilters}', ignoring`
      );
      return;
    }

    for (const filter of result.filters) {
      const storedIndex = this.filters.values.findIndex(
        (f) => f.raw === filter.text
      );
      this.filters.values[storedIndex] = Object.assign(
        this.filters.values[storedIndex],
        {
          applied: true,
          isValid: filter.isValid,
          hits: filter.hits,
          name: filter.name,
          matcher: filter.matcher,
          value: filter.value,
        }
      );
    }

    let updates = {};
    for (const key of [
      "colors",
      "counters",
      "grids",
      "silences",
      "upstreams",
      "receivers",
    ]) {
      updates[key] = result[key];
    }
    this.data = Object.assign(this.data, updates);

    // before storing new version check if we need to reload
    if (
      this.info.version !== "unknown" &&
      this.info.version !== result.version
    ) {
      this.info.upgradeNeeded = true;
    }
    // update extra root level keys that are stored under 'info'
    for (const key of ["totalAlerts", "version", "authentication"]) {
      if (this.info[key] !== result[key]) {
        this.info[key] = result[key];
      }
    }

    // settings exported via API
    this.settings.values = result.settings;

    this.status.setIdle();
  });

  handleFetchError = action((err) => {
    this.status.setFailure(err);

    // reset alert counter since we won't be rendering any alerts
    this.info.totalAlerts = 0;

    // all unapplied filters should be marked applied to reset progress indicator
    this.filters.applyAllFilters();

    return { error: err };
  });
}

export {
  AlertStore,
  AlertStoreStatuses,
  FormatBackendURI,
  FormatAPIFilterQuery,
  FormatAlertsQ,
  DecodeLocationSearch,
  UpdateLocationSearch,
  NewUnappliedFilter,
};
