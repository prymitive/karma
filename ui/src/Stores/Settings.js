import { action } from "mobx";
import { localStored } from "mobx-stored";

class SavedFilters {
  config = localStored(
    "savedFilters",
    {
      filters: [],
      present: false
    },
    {
      delay: 100
    }
  );

  save = action(newFilters => {
    this.config.filters = newFilters;
    this.config.present = true;
  });

  clear = action(() => {
    this.config.filters = [];
    this.config.present = false;
  });
}

class FetchConfig {
  config = localStored("fetchConfig", { interval: 30 }, { delay: 100 });

  setInterval = action(newInterval => {
    this.config.interval = newInterval;
  });
}

class AlertGroupConfig {
  options = Object.freeze({
    expanded: { label: "Always expanded", value: "expanded" },
    collapsedOnMobile: {
      label: "Collapse on mobile",
      value: "collapsedOnMobile"
    },
    collapsed: { label: "Always collapsed", value: "collapsed" }
  });
  config = localStored(
    "alertGroupConfig",
    {
      defaultRenderCount: 5,
      defaultCollapseState: this.options.collapsedOnMobile.value,
      colorTitleBar: false
    },
    { delay: 100 }
  );

  update = action(data => {
    for (const [key, val] of Object.entries(data)) {
      this.config[key] = val;
    }
  });
}

class SilenceFormConfig {
  config = localStored("silenceFormConfig", { author: "" }, { delay: 100 });

  saveAuthor = action(newAuthor => {
    this.config.author = newAuthor;
  });
}

class GridConfig {
  options = Object.freeze({
    default: { label: "Use defaults from karma config file", value: "default" },
    disabled: { label: "No sorting", value: "disabled" },
    startsAt: { label: "Sort by alert timestamp", value: "startsAt" },
    label: { label: "Sort by alert label", value: "label" }
  });
  config = localStored(
    "alertGridConfig",
    {
      sortOrder: this.options.default.value,
      groupWidth: 420
    },
    { delay: 100 }
  );
}

class FilterBarConfig {
  config = localStored(
    "filterBarConfig",
    {
      autohide: true
    },
    {
      delay: 100
    }
  );
}

class TopLabelsConfig {
  config = localStored(
    "topLabelsConfig",
    {
      show: true,
      minPercent: 33
    },
    {
      delay: 100
    }
  );
}

class Settings {
  constructor() {
    this.savedFilters = new SavedFilters();
    this.fetchConfig = new FetchConfig();
    this.alertGroupConfig = new AlertGroupConfig();
    this.gridConfig = new GridConfig();
    this.silenceFormConfig = new SilenceFormConfig();
    this.filterBarConfig = new FilterBarConfig();
    this.topLabelsConfig = new TopLabelsConfig();
  }
}

export { Settings };
