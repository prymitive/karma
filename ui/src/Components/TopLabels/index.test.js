import React from "react";

import { Provider } from "mobx-react";

import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";

import { TopLabels } from ".";

let alertStore;
let settingsStore;

beforeEach(() => {
  alertStore = new AlertStore([]);
  settingsStore = new Settings();
});

const MountedTopLabels = () => {
  return mount(
    <Provider alertStore={alertStore}>
      <TopLabels alertStore={alertStore} settingsStore={settingsStore} />
    </Provider>
  );
};

describe("<TopLabels />", () => {
  it("doesn't render anything when settingsStore.topLabelsConfig.config.show is false", () => {
    settingsStore.topLabelsConfig.config.show = false;
    const tree = MountedTopLabels();
    expect(tree.html()).toBe("");
  });

  it("matches snapshot", () => {
    settingsStore.topLabelsConfig.config.show = true;
    alertStore.data.counters = [{ name: "foo", value: "bar", percent: "50" }];
    const tree = MountedTopLabels();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("doesn't render labels with 100%", () => {
    settingsStore.topLabelsConfig.config.show = true;
    settingsStore.topLabelsConfig.config.minPercent = 1;
    alertStore.data.counters = [
      { name: "foo80", value: "80", percent: 80 },
      { name: "foo100", value: "100", percent: 100 },
      { name: "foo74", value: "74", percent: 74 }
    ];
    const tree = MountedTopLabels();
    expect(tree.find(".components-label")).toHaveLength(2);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("renders only labels with >=75% when settingsStore.topLabelsConfig.config.minPercent is 75", () => {
    settingsStore.topLabelsConfig.config.show = true;
    settingsStore.topLabelsConfig.config.minPercent = 75;
    alertStore.data.counters = [
      { name: "foo80", value: "80", percent: 80 },
      { name: "foo70", value: "70", percent: 70 },
      { name: "foo74", value: "74", percent: 74 },
      { name: "foo75", value: "75", percent: 75 },
      { name: "foo15", value: "15", percent: 15 },
      { name: "foo99", value: "99", percent: 99 }
    ];
    const tree = MountedTopLabels();
    expect(tree.find(".components-label")).toHaveLength(3);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("clicking on the close icon sets settingsStore.topLabelsConfig.config.show to 'false'", () => {
    settingsStore.topLabelsConfig.config.show = true;
    alertStore.data.counters = [{ name: "foo", value: "99", percent: 99 }];
    const tree = MountedTopLabels();
    tree.find("svg.close").simulate("click");
    expect(settingsStore.topLabelsConfig.config.show).toBe(false);
  });
});
