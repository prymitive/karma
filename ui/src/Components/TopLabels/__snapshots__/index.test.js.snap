// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopLabels /> doesn't render labels with 100% 1`] = `
"
<div class=\\"components-toplabels m-1 bg-secondary rounded d-flex flex-row\\">
  <div class=\\"flex-grow-0 flex-shrink-0 pl-1 d-none d-sm-block\\">
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-3\\"
         data-original-title=\\"Top labels\\"
    >
      <span>
        <svg aria-hidden=\\"true\\"
             focusable=\\"false\\"
             data-prefix=\\"far\\"
             data-icon=\\"chart-bar\\"
             class=\\"svg-inline--fa fa-chart-bar fa-w-16 text-white\\"
             role=\\"img\\"
             xmlns=\\"http://www.w3.org/2000/svg\\"
             viewbox=\\"0 0 512 512\\"
        >
          <path fill=\\"currentColor\\"
                d=\\"M396.8 352h22.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-192 0h22.4c6.4 0 12.8-6.4 12.8-12.8V140.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h22.4c6.4 0 12.8-6.4 12.8-12.8V204.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zM496 400H48V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16zm-387.2-48h22.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8z\\"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
  <div class=\\"pl-1 flex-grow-1 flex-shrink-1 components-toplabels-labels\\">
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-4\\"
         data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
    >
      <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover pl-0 text-left\\">
        <span class=\\"mr-1 px-1 bg-light text-primary components-toplabel-percent\\">
          80%
        </span>
        <span class=\\"components-label-name\\">
          foo80:
        </span>
        <span class=\\"components-label-value\\">
          80
        </span>
      </span>
    </div>
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-5\\"
         data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
    >
      <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover pl-0 text-left\\">
        <span class=\\"mr-1 px-1 bg-light text-primary components-toplabel-percent\\">
          74%
        </span>
        <span class=\\"components-label-name\\">
          foo74:
        </span>
        <span class=\\"components-label-value\\">
          74
        </span>
      </span>
    </div>
  </div>
  <div class=\\"flex-grow-0 flex-shrink-0 px-1\\">
    <span>
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"times\\"
           class=\\"svg-inline--fa fa-times fa-w-11 cursor-pointer close\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 352 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"
        >
        </path>
      </svg>
    </span>
  </div>
</div>
"
`;

exports[`<TopLabels /> matches snapshot 1`] = `
"
<div class=\\"components-toplabels m-1 bg-secondary rounded d-flex flex-row\\">
  <div class=\\"flex-grow-0 flex-shrink-0 pl-1 d-none d-sm-block\\">
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-1\\"
         data-original-title=\\"Top labels\\"
    >
      <span>
        <svg aria-hidden=\\"true\\"
             focusable=\\"false\\"
             data-prefix=\\"far\\"
             data-icon=\\"chart-bar\\"
             class=\\"svg-inline--fa fa-chart-bar fa-w-16 text-white\\"
             role=\\"img\\"
             xmlns=\\"http://www.w3.org/2000/svg\\"
             viewbox=\\"0 0 512 512\\"
        >
          <path fill=\\"currentColor\\"
                d=\\"M396.8 352h22.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-192 0h22.4c6.4 0 12.8-6.4 12.8-12.8V140.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h22.4c6.4 0 12.8-6.4 12.8-12.8V204.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zM496 400H48V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16zm-387.2-48h22.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8z\\"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
  <div class=\\"pl-1 flex-grow-1 flex-shrink-1 components-toplabels-labels\\">
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-2\\"
         data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
    >
      <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover pl-0 text-left\\">
        <span class=\\"mr-1 px-1 bg-light text-primary components-toplabel-percent\\">
          50%
        </span>
        <span class=\\"components-label-name\\">
          foo:
        </span>
        <span class=\\"components-label-value\\">
          bar
        </span>
      </span>
    </div>
  </div>
  <div class=\\"flex-grow-0 flex-shrink-0 px-1\\">
    <span>
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"times\\"
           class=\\"svg-inline--fa fa-times fa-w-11 cursor-pointer close\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 352 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"
        >
        </path>
      </svg>
    </span>
  </div>
</div>
"
`;

exports[`<TopLabels /> renders only labels with >=75% when settingsStore.topLabelsConfig.config.minPercent is 75 1`] = `
"
<div class=\\"components-toplabels m-1 bg-secondary rounded d-flex flex-row\\">
  <div class=\\"flex-grow-0 flex-shrink-0 pl-1 d-none d-sm-block\\">
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-6\\"
         data-original-title=\\"Top labels\\"
    >
      <span>
        <svg aria-hidden=\\"true\\"
             focusable=\\"false\\"
             data-prefix=\\"far\\"
             data-icon=\\"chart-bar\\"
             class=\\"svg-inline--fa fa-chart-bar fa-w-16 text-white\\"
             role=\\"img\\"
             xmlns=\\"http://www.w3.org/2000/svg\\"
             viewbox=\\"0 0 512 512\\"
        >
          <path fill=\\"currentColor\\"
                d=\\"M396.8 352h22.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-192 0h22.4c6.4 0 12.8-6.4 12.8-12.8V140.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h22.4c6.4 0 12.8-6.4 12.8-12.8V204.8c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zM496 400H48V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-16c0-8.84-7.16-16-16-16zm-387.2-48h22.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-22.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8z\\"
          >
          </path>
        </svg>
      </span>
    </div>
  </div>
  <div class=\\"pl-1 flex-grow-1 flex-shrink-1 components-toplabels-labels\\">
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-7\\"
         data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
    >
      <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover pl-0 text-left\\">
        <span class=\\"mr-1 px-1 bg-light text-primary components-toplabel-percent\\">
          80%
        </span>
        <span class=\\"components-label-name\\">
          foo80:
        </span>
        <span class=\\"components-label-value\\">
          80
        </span>
      </span>
    </div>
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-8\\"
         data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
    >
      <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover pl-0 text-left\\">
        <span class=\\"mr-1 px-1 bg-light text-primary components-toplabel-percent\\">
          75%
        </span>
        <span class=\\"components-label-name\\">
          foo75:
        </span>
        <span class=\\"components-label-value\\">
          75
        </span>
      </span>
    </div>
    <div class
         style=\\"display: inline-block; max-width: 100%;\\"
         data-tooltipped
         aria-describedby=\\"tippy-tooltip-9\\"
         data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
    >
      <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover pl-0 text-left\\">
        <span class=\\"mr-1 px-1 bg-light text-primary components-toplabel-percent\\">
          99%
        </span>
        <span class=\\"components-label-name\\">
          foo99:
        </span>
        <span class=\\"components-label-value\\">
          99
        </span>
      </span>
    </div>
  </div>
  <div class=\\"flex-grow-0 flex-shrink-0 px-1\\">
    <span>
      <svg aria-hidden=\\"true\\"
           focusable=\\"false\\"
           data-prefix=\\"fas\\"
           data-icon=\\"times\\"
           class=\\"svg-inline--fa fa-times fa-w-11 cursor-pointer close\\"
           role=\\"img\\"
           xmlns=\\"http://www.w3.org/2000/svg\\"
           viewbox=\\"0 0 352 512\\"
      >
        <path fill=\\"currentColor\\"
              d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"
        >
        </path>
      </svg>
    </span>
  </div>
</div>
"
`;
